overlord/managers_test.go

```diff
@@ -714,6 +795,340 @@ apps:
+}
+
+func (ms *mgrsSuite) TestHappyRemoteInstallAndUpdateWithPostHocEpochBump(c *C) {
+	// test install through store and update, where there is an epoch
+	// bump in the upgrade that comes in after the initial update is
+	// computed.
+
+	// this is mostly checking the same as TestHappyRemoteInstallAndUpdateWithEpochBump
+	// but serves as a sanity check for the Without case that follows
+	// (these two together serve as a test for the refresh filtering)
+	ms.testHappyRemoteInstallAndUpdateWithMaybeEpochBump(c, true)
+}
+
+func (ms *mgrsSuite) TestHappyRemoteInstallAndUpdateWithoutEpochBump(c *C) {
+	// test install through store and update, where there _isn't_ an epoch bump in the upgrade
+	// note that there _are_ refreshes available after the refresh,
+	// but they're not an epoch bump so they're ignored
+	ms.testHappyRemoteInstallAndUpdateWithMaybeEpochBump(c, false)
+}
+
+func (ms *mgrsSuite) testHappyRemoteInstallAndUpdateWithMaybeEpochBump(c *C, doBump bool) {
+	ms.prereqSnapAssertions(c)
+
+	snapPath, _ := ms.makeStoreTestSnap(c, "{name: foo, version: 1}", "1")
+	ms.serveSnap(snapPath, "1")
+
+	mockServer := ms.mockStore(c)
+	defer mockServer.Close()
+
+	st := ms.o.State()
+	st.Lock()
+	defer st.Unlock()
+
+	ts, err := snapstate.Install(st, "foo", "stable", snap.R(0), 0, snapstate.Flags{})
+	c.Assert(err, IsNil)
+	chg := st.NewChange("install-snap", "...")
+	chg.AddAll(ts)
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	// confirm it worked
+	c.Assert(chg.Status(), Equals, state.DoneStatus, Commentf("install-snap change failed with: %v", chg.Err()))
+
+	// sanity checks
+	info, err := snapstate.CurrentInfo(st, "foo")
+	c.Assert(err, IsNil)
+	c.Assert(info.Revision, Equals, snap.R(1))
+	c.Assert(info.SnapID, Equals, fooSnapID)
+	c.Assert(info.Epoch.String(), Equals, "0")
+
+	// add a new revision
+	snapPath, _ = ms.makeStoreTestSnap(c, "{name: foo, version: 2}", "2")
+	ms.serveSnap(snapPath, "2")
+
+	// refresh
+
+	ts, err = snapstate.Update(st, "foo", "stable", snap.R(0), 0, snapstate.Flags{})
+	c.Assert(err, IsNil)
+	chg = st.NewChange("upgrade-snap", "...")
+	chg.AddAll(ts)
+
+	// add another new revision, after the update was computed (maybe with an epoch bump)
+	if doBump {
+		snapPath, _ = ms.makeStoreTestSnap(c, "{name: foo, version: 3, epoch: 1*}", "3")
+	} else {
+		snapPath, _ = ms.makeStoreTestSnap(c, "{name: foo, version: 3}", "3")
+	}
+	ms.serveSnap(snapPath, "3")
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(chg.Err(), IsNil)
+	c.Assert(chg.Status(), Equals, state.DoneStatus, Commentf("upgrade-snap change failed with: %v", chg.Err()))
+
+	info, err = snapstate.CurrentInfo(st, "foo")
+	c.Assert(err, IsNil)
+
+	if doBump {
+		// if the epoch bumped, then we should've re-refreshed
+		c.Check(info.Revision, Equals, snap.R(3))
+		c.Check(info.SnapID, Equals, fooSnapID)
+		c.Check(info.Epoch.String(), Equals, "1*")
```

Comment:
Would it make sense to check the change here that it actually installed foo twice, once in revno 2 and once in revno 3? To ensure that the snap did not go directly to revno(3) and we did in fact have a re-refresh?